### ISSUE TITLE
'fig up' should return with non-zero exit-code if some container "failed"

### ISSUE THREAD

**Issue by matleh** [docker/compose#683]
`fig run <container>` does already return the exit-code of the _container_ (see #197) but `fig up` returns with 0 no matter what the exit-code of the started containers is.

It would be helpful to able to detect, if one of the containers returned with a non-zero exit code and have fig return a non-zero exit-code in that case, too.

Also, `fig up` does end with exit-code 1 if one of the containers could not be started at all, and also if one does _Ctl-C_ once, but it ends with exit-code 0 of one does _Ctl-C_ twice. It would be more in line with "standard unix behavior" if fig would exit with -2 in case of single _Ctl-C_ (SIGINT) and -9 in case of double _Ctl-C_ (SIGKILL).

**Comment by dvapelnik** [comment]
I think that fig must return more meaningful exit-codes
I'm writing bash wrapper for **fig** service with deploying database dump from file and other things, but I want to check each my step for rollbacking all my actions for prevent corrupting my data (database dump by example)
So, by example, my steps:

1. Start up containers *(I'm checking exit code and going next step if it equal to zero)*
2. //todo something with database and DNS
3. I'm stopping my containers
4. I'm removing my containers

In this workflow all look good, but if my heedless user accidentally run this steps:

1. Start up containers *(I'm checking exit code and going next step if it equal to zero)*
2. //todo something with database and DNS
3. Start up containers again with ```fig up -d``` *(I'm checking exit code and exit code is zero even with using ```--no-recreate``` command line key)*
4. My wrapper will deploy database from dump-file again with loss t my actual data in database

**Can I check is containers are running?** I can get ```fig ps``` and grep it but I think that is the wrong way 

I thing that situation with workflow is wrong because ```fig up``` must only start up containers. I'll ```fig stop```, ```fig rm``` and ```fig up``` again if i want to recreate my containers. I think that is a correct workflow with **fig**. I'm expecting nonzero exit code on ```fig up``` when containers are running but I look unexpected zero exit code.

**Comment by smecsia** [comment]
I've found the workaround of how to catch the exit code in fig. This is not very clean and nice solution, but at least it works for me.
```
exec 5>&1
log=$(sudo fig up | tee /dev/fd/5)
exitcode=$(echo $log | grep 'exited with code' | sed 's/^.*exited with code \([0-9]\+\).*$/\1/g')
```

**Comment by vvitayau** [comment]
FIG_SERVICE=test
DOCKER_CONTAINER=`fig -p xxx ps -q $FIG_SERVICE | cut -c -12`
docker ps -a --filter 'exited=0' | grep $DOCKER_CONTAINER
EXIT_CODE=$?

**Comment by johanhaleby** [comment]
+1

**Comment by ebuchman** [comment]
Just got tripped up by this as well. Using `fig up` for continuous integration on circle ci. My `fig.yml` simply calls a `test.sh` file.  So my solution (courtesy of @xcthulhu) was to add a `touch /tmp/success` at the end of `test.sh` and to check for it with `test -e /tmp/success` at the end of the `circle.yml`. Worked beautifully, and much prettier than solutions above if your setup is similar.

**Comment by dnephin** [comment]
Why not use `fig run tester` instead? You'll get the exit code you expect.

**Comment by MrMMorris** [comment]
@ebuchman I am in the exact same situation. I will try your workaround. 

But this definitely needs to be handled properly in compose

**Comment by MrMMorris** [comment]
@dnephin it seems your suggestion doesn't work? Am I doing something wrong?

docker-compose.yml
```
test:
  image: node:latest
  command: sleep 100
```

```
$ docker-compose up -d
Creating core_test_1...

$ docker-compose ps
   Name        Command    State   Ports
---------------------------------------
core_test_1   sleep 100   Up

$ docker-compose run test npm test
> tv-core@1.0.0 pretest /app
> jshint .

< LOTS OF LINT ERRORS>

> ERR! Test failed.  See above for more details.

$ echo $!
0
```

**Comment by dnephin** [comment]
> I feel it is just another workaround for what should work as expected

I'm not really convinced of that. I would much rather see `docker-compose up` go the way of #741. The exit status of `docker-compose up` only reflects "docker-compose was able to **start** containers".  At that point `up` is done it's work, and what you're really seeing is the logs. If at any point it fails, it is not related to the original `docker-compose up` you ran. Having `docker-compose up` attempt to reflect the many different possible exit scenarios is just not realistic. 

On the other hand, `docker-compose run` has the behaviour of running one specific container (with its dependencies), so it's clear that when that container exits the exit status of `docker-compose run`  would match that single container.

>  I use separate test containers that have all my test framework dependancies so I can just run them 'attached' them to my app containers and get test results.

I don't understand what you're trying to do here. Could you link to a sample config?

I can say that my primary (and almost exclusive) use-case for docker-compose is CI, and I have never encountered any issue with this.

```
docker-compose run tester my_tests
```

**Comment by dnephin** [comment]
That should be `echo $?`

This works:

**docker-compose.yml**

```yaml
tester:
    image: busybox
    command: sleep 100
```

```bash
$ docker-compose run tester cat /file/does/not/exist
cat: can't open '/file/does/not/exist': No such file or directory
$ echo $?
1
```

**Comment by MrMMorris** [comment]
Yep, that works. Not sure where I got `$!` from...

Your workaround is actually fine for me. I understand the complications of handling exited codes from process in a container that don't relate to `fig up`. I also think that this would be something most people expect and it kind of sucks for a container to fail immediately after starting and not know about it....

Either way I am fine with just using `fig run` for now.

**Comment by spenthil** [comment]
Alternatively:

`docker-compose ps -q | xargs docker inspect -f '{{ .State.ExitCode }}' | grep -v 0 | wc -l | tr -d ' '`

1. get container IDs
2. get last runs exit code for each container ID
3. only non-0 status codes
4. count number of non-0 status codes
5. trim out white space

Returns how many non-0 exit codes were returned. Would be 0 if everything exited with code 0.

**Comment by sanmai-NL** [comment]
Using the `up` subcommand to start container with entry point `/bin/sh` and command `exit 40` prints to the console:
`my_container exited with code 0`

**Comment by dnephin** [comment]
Fixed in #4397 ` --exit-code-from`
